src/TLuaInterpreter.cpp

```diff
@@ -10739,6 +10750,505 @@ int TLuaInterpreter::restartIrc(lua_State* L)
+        lua_pushstring(L, "ttsSetVoiceByIndex: voice index out of bounds");
+        lua_error(L);
+        return 1;
+    }
+    
+    speechUnit->setVoice(speechVoices.at(index));
+    
+    return 0;
+}
+
+/**
+ *  Handles playback of queued text.
+ */
+void TLuaInterpreter::ttsStateChanged(QTextToSpeech::State state)
+{
+    TEvent event;
+    switch(state)
+    {
+        case QTextToSpeech::Paused : event.mArgumentList.append(QLatin1String("ttsSpeechPaused")); break;
+        case QTextToSpeech::Speaking : event.mArgumentList.append(QLatin1String("ttsSpeechStarted")); break;
+        case QTextToSpeech::BackendError : event.mArgumentList.append(QLatin1String("ttsSpeechError")); break;
+        case QTextToSpeech::Ready : event.mArgumentList.append(QLatin1String("ttsSpeechReady")); break;
+    }
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    
+    if (state == QTextToSpeech::Speaking)
+    {
+        event.mArgumentList.append(speechCurrent);
+        event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    }
+    
+    mudlet::self()->getHostManager().postInterHostEvent(NULL, event, true);
+    
+    if(state != QTextToSpeech::Ready || speechQueue.empty())
+    {
+        return;
+    }
+    
+    speechCurrent = speechQueue.takeFirst();
+    
+    speechUnit->say(speechCurrent);
+    
+    return;
+}
+
+
+/** ttsQueueSpeech( input, opt. index )
+ *  Adds a line of text to the queue at the specified index.
+ *  If no index is provided, it will add it to the end of the queue.
+ */
+int TLuaInterpreter::ttsQueueSpeech(lua_State* L)
+{
+    TLuaInterpreter::ttsBuild();
+    
+    if(!lua_isstring(L, 1)) {
+        lua_pushfstring(L, "ttsQueueText: bad argument #1 type (input as string expected, got %s!)", luaL_typename(L, 1));
+        lua_error(L);
+        return 1;
+    }
+    
+    QString inputText = lua_tostring(L, 1);
+    int index;
+    
+    if(lua_gettop(L) > 1) {
+        if(!lua_isnumber(L, 2)) {
+            lua_pushfstring(L, "ttsQueueText: bad argument #2 type (optional index as number expected, got %s!)", luaL_typename(L, 1));
+            lua_error(L);
+            return 1;
+        }
+        
+        index = lua_tonumber(L, 2);
+        index--;
+        
+        if(index < 0) {
+            index = 0;
+		}
+        
+		if(index > speechQueue.size()) {
+            index = speechQueue.size();
+		}
+    }
+    else {
+        index = speechQueue.size();
+    }
+
+    speechQueue.insert(index, inputText);
+    
+    TEvent event;
+    Host& host = getHostFromLua(L);
+    event.mArgumentList.append(QLatin1String("ttsSpeechQueued"));
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    event.mArgumentList.append(inputText);
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_STRING);
+    event.mArgumentList.append(QString::number(index));
+    event.mArgumentTypeList.append(ARGUMENT_TYPE_NUMBER);
+    host.raiseEvent(event);
+    
+    TLuaInterpreter::ttsStateChanged(speechUnit->state());
```

Comment:
You did not check if the state did change. If something was being said and you queue another line, the state of "started" won't have changed. Effectively you always send 2 events in a row.